lib/Runtime/Library/JavascriptArray.cpp

```diff
@@ -9152,44 +9152,63 @@ namespace Js
             uint32 fromIndex = static_cast<uint32>(fromVal);
             uint32 toIndex = static_cast<uint32>(toVal);
 
-            while (count > 0)
+            if (typedArrayBase && length <= typedArrayBase->GetLength())
             {
-                JS_REENTRANT(jsReentLock, BOOL hasItem = obj->HasItem(fromIndex));
-                if (hasItem)
+                // Typed array can't suddenly transform into a different type like JavascriptArray can,
+                // and can't contain holes, so as long as `length` is no more than [[ArrayLength]], we can
+                // ignore [[HasProperty]] and DeletePropertyOrThrow steps.
+                while (count > 0)
                 {
-                    if (typedArrayBase)
-                    {
-                        Var val = typedArrayBase->DirectGetItem(fromIndex);
+                    Var val = typedArrayBase->DirectGetItem(fromIndex);
 
-                        JS_REENTRANT(jsReentLock, typedArrayBase->DirectSetItem(toIndex, val));
-                    }
-                    else if (pArr)
+                    JS_REENTRANT(jsReentLock, typedArrayBase->DirectSetItem(toIndex, val));
```

Comment:
>val [](start = 85, length = 3)

How much benefit do we get from this fast path?  If you don't much, then probably don't want to add to the complexity.

If you are doing this optimization, you might want to do it fully, by avoiding a virtual call to DirectGetItem/DirectSetItem on every iteration of the loop, and have a specialized implementation of these that will memcpy if you are going forward, and loop copy if you going backward, instead of converting all the values to "Var"s.   That might getting a more notifiable  benefit then what you have here.

Reply:
It roughly doubles the speed of a large copyWithin operation on my machine. Agreed, we can probably speed up by another order of magnitude using straightforward memory operations.